src/sagemaker_xgboost_container/serving_mms.py

```diff
@@ -0,0 +1,121 @@
+    # sometimes fail with 'bad address'. more investigation needed
+    # retry starting mms until it's ready
+    logging.info("Trying to set up model server handler: {}".format(handler))
+    _set_mms_configs(is_multi_model, handler)
+    model_server.start_model_server(handler_service=handler,
+                                    is_multi_model=is_multi_model,
+                                    config_file=get_mms_config_file_path())
+
+
+def _is_multi_model_endpoint():
+    if "SAGEMAKER_MULTI_MODEL" in os.environ and os.environ["SAGEMAKER_MULTI_MODEL"] == 'true':
+        return True
+    else:
+        return False
+
+
+def _set_mms_configs(is_multi_model, handler):
+    """Set environment variables for MMS to parse during server initialization. These env vars are used to
+    propagate the config.properties file used during MxNet Model Server initialization.
+    'SAGEMAKER_MMS_MODEL_STORE' has to be set to the model location during single model inference because MMS
+    is initialized with the model. In multi-model mode, MMS is started with no models loaded.
+    Note: Ideally, instead of relying on env vars, this should be written directly to a config file.
+    """
+    max_content_length = os.getenv("MAX_CONTENT_LENGTH", DEFAULT_MAX_CONTENT_LEN)
+    if int(max_content_length) > MAX_CONTENT_LEN_LIMIT:
+        # Cap at 20mb
+        max_content_length = MAX_CONTENT_LEN_LIMIT
+
+    max_workers = multiprocessing.cpu_count()
+    max_job_queue_size = 2 * max_workers
+
+    # Max heap size = (max workers + max job queue size) * max payload size * 1.2 (20% buffer) + 128 (base amount)
+    max_heap_size = ceil((max_workers + max_job_queue_size) * (int(max_content_length) / 1024 ** 2) * 1.2) + 128
+
+    os.environ["SAGEMAKER_NUM_MODEL_WORKERS"] = '1'
+    os.environ["SAGEMAKER_MODEL_JOB_QUEUE_SIZE"] = '2'
+    os.environ["SAGEMAKER_MMS_MODEL_STORE"] = '/'
+    os.environ["SAGEMAKER_MMS_LOAD_MODELS"] = ''
+
+    if not os.getenv("SAGEMAKER_BIND_TO_PORT", None):
+        os.environ["SAGEMAKER_BIND_TO_PORT"] = str(PORT)
+
+    os.environ["SAGEMAKER_MAX_HEAP_SIZE"] = str(max_heap_size) + 'm'
+    os.environ["SAGEMAKER_MAX_DIRECT_MEMORY_SIZE"] = os.environ["SAGEMAKER_MAX_HEAP_SIZE"]
+    os.environ["SAGEMAKER_MAX_REQUEST_SIZE"] = str(max_content_length)
+    os.environ["SAGEMAKER_MMS_DEFAULT_HANDLER"] = handler
+
+    MMS_CONFIG_FILE_PATH = get_mms_config_file_path()
+
+    # TODO: Revert config.properties.tmp to config.properties and add back in vmargs
+    # set with environment variables after MMS implements parsing environment variables
+    # for vmargs, update MMS section of final/Dockerfile.cpu to match, and remove the
+    # following code.
+    try:
+        with open(MMS_CONFIG_FILE_PATH + '.tmp', 'r') as f:
+            with open(MMS_CONFIG_FILE_PATH, 'w+') as g:
+                g.write("vmargs=-XX:-UseLargePages -XX:+UseG1GC -XX:MaxMetaspaceSize=32M -XX:+ExitOnOutOfMemoryError "
```

Comment:
Using ExitOnOutOfMemoryError might result in JVM abruptly getting killed when OOM occurs. This could be bad because a sudden spike of traffic will render the host unusable. Consider having it / removing it based on the promised contract exposed to the user.